src/esi/Esi.cc

```diff
@@ -1256,8 +1256,17 @@ ESIContext::parse()
 
         PROF_start(esiParsing);
 
-        while (buffered.getRaw() && !flags.error)
-            parseOneBuffer();
+        try {
+            while (buffered.getRaw() && !flags.error)
+                parseOneBuffer();
+
+        } catch (const TextException &ex) {
```

Comment:
Please merge the two catch clauses into one and use CurrentException to set the error message: See SquidMainSafe() and StoreEntry::checkDisk() for catching examples. And ToSBuf() can be used to convert some error message prefix and CurrentException into SBuf.

Reply:
I think you have missed the purpose of these. If we use the unified catch and re-throw like those examples then any malformed ESI response will halt the entire Squid process.

Catching the TextException is done to convert _only_ the errors caught by our parse code into ESI error pages. All other exceptions are more serious and need to be re-thrown for that more fatal handling by SquidMainSafe() itself.

Reply:
> If we use the unified catch and re-throw ...

I did not suggest doing that. I suggested that we use a "unified catch". See SquidMainSafe() and StoreEntry::checkDisk() for _catching_ examples, not for (re)throwing examples. Rethrowing is what happens after the exception is caught. It is a part of exception handling. ESI exception handling needs are different from those of SquidMainSafe() and StoreEntry::checkDisk().

There should be no rethrow in this ESI context at all AFAICT -- we handle all exceptions by (terminating the parsing loop and) setting the error message. The catch-and-rethrow clause in the current (and reviewed earlier) PR code does nothing -- it should not be there at all (in its current shape). Sorry I did not say that explicitly. I wanted to stay positive.

Rule of thumb: Do not catch exceptions that you do not intend to (at least partially) handle.

Reply:
These TextException thrown by the parser are being placed into error pages displayed to the client. If we catch all exceptions and display the text from system exceptions we will be displaying machine internal address locations and OS internal details in an essentially public output.

Reply:
> If we catch all exceptions and display the text from system exceptions...

There are no "system exceptions" and "not-system exceptions" in Squid exception types. There are also no "sensitive exceptions" and "not-sensitive exceptions" in Squid exception types. This PR does not protect Squid from leaking sensitive info by treating exceptions of type TextExceptions differently from other exception types.

Today, there is no valid reason to treat TextExceptions differently than all other exception types in this context. And there is virtually never a valid reason to add code that does nothing. My change request fixes both problems (without changing Squid security posture).

We can, of course, discuss whether or how Squid exception throwing and/or catching code should protect sensitive info, but that discussion is not specific to this particular code and, most likely, should not happen in this PR.

Reply:
The ESI parsing includes XML parsing sub-calls to libxml2 or libexpat. They include sub-calls to stdlib for memory allocations and other such things which may trigger; for example an alloc failure exception - which contains address locations for memory blocks.
 If we catch such exceptions and transmit in error page texts to remote hosts that is a leak of ASLR information - which is very sensitive and immediate CVE assignment when spotted.

Since those exceptions are coming from third-party code and/or transitory system activity we cannot assume their absence or safety. Only our TextException messages are certain not to have any sensitive information (because I/we have checked for that) - and if they do that is a bug we can fix without loosing the needed info.

Reply:
> Only our TextException messages are certain not to have any sensitive information (because I/we have checked for that)

There are lower-level mistakes in your response, but I will try to stay focused on the big picture in hope to reduce time waste: "I have checked" arguments about specific TextException messages cannot address this change request because it is not about specific exceptions. C++ exception catchers are based on types, not individual exception objects. No mortal, not even you, can "check" what specific exception objects this code will catch tomorrow. And as far as types are concerned, the TextException class, just like std::runtime_error, is not designed to be "not-sensitive" -- TextException objects may contain sensitive information.

There are several valid ways to address this change request, including these two:

1. Treat all exceptions the same. Do not copy their specifics to the response (because those specifics might contain sensitive info and because they are not really meant/designed for sharing with end-users). Report all caught exceptions to cache.log (with a rate or count limit for level-1 reports), where a capable admin can find them.
2. Allow throwers to mark some exceptions as "safe to share with end users" (or something like that). This can be done using a general mechanism (to be reused in other contexts) or something designed specifically for ESI parsing (can carry context-specific but still non-sensitive info). Add marked exceptions to the response. Report all (other?) exceptions to cache.log (with a rate or count limit for level-1 reports), where a capable admin can find them.

I recommend the first option because it is simpler and focuses on the primary problem this PR is meant to solve.

Please let me know if you need my help with fleshing out one of the above options.